File: tools/process_manifest.py
import argparse
import json

parser = argparse.ArgumentParser(
    description="Build manifest files from 'mc_manifest.json'."
)
parser.add_argument(
    "--target",
    choices=["release", "debug", "server"],
    default="debug",
    help="Whether to build the addon in debug or release mode or for servers.",
)
args = parser.parse_args()

bp_manifest = {}
rp_manifest = {}


def processJsonElement(element, bp_element, rp_element):
    def process(key, value):
        if isinstance(value, dict):
            bp_element[key] = {}
            rp_element[key] = {}
            processJsonElement(value, bp_element[key], rp_element[key])
        elif isinstance(value, list):
            bp_element[key] = []
            rp_element[key] = []
            processJsonElement(value, bp_element[key], rp_element[key])
        else:
            if isinstance(bp_element, list):
                bp_element.append(value)
                rp_element.append(value)
            else:
                bp_element[key] = value
                rp_element[key] = value

    if isinstance(element, dict):
        for [key, value] in element.items():
            if key.startswith("bp_"):
                if key.startswith("bp_server_"):
                    sub = bp_element[key[10:]]
                    if isinstance(sub, list):
                        bp_element[key[10:]] = sub + value
                    elif isinstance(sub, dict):
                        bp_element[key[10:]] = {**sub, **value}
                else:
                    bp_element[key[3:]] = value
            elif key.startswith("rp_"):
                rp_element[key[3:]] = value
            else:
                process(key, value)
    elif isinstance(element, list):
        i = 0
        for value in element:
            process(i, value)
            i = i + 1


# load base manifest
with open("mc_manifest.json", "r") as file:
    manifest = json.load(file)
    processJsonElement(manifest, bp_manifest, rp_manifest)

version = manifest["header"]["version"]
bp_manifest["header"]["name"] += " " + ".".join(map(str, version))
rp_manifest["header"]["name"] += " " + ".".join(map(str, version))

if not type(version) is str:
    version = version[:3]
bp_manifest["header"]["version"] = version
rp_manifest["header"]["version"] = version

if not "dependencies" in bp_manifest:
    bp_manifest["dependencies"] = []
bp_manifest["dependencies"].append(
    {"uuid": rp_manifest["header"]["uuid"], "version": rp_manifest["header"]["version"]}
)

if not "dependencies" in rp_manifest:
    rp_manifest["dependencies"] = []
rp_manifest["dependencies"].append(
    {"uuid": bp_manifest["header"]["uuid"], "version": bp_manifest["header"]["version"]}
)

if args.target == "debug":
    bp_manifest["header"]["name"] += " [DEBUG]"
    rp_manifest["header"]["name"] += " [DEBUG]"

# export behaviour and resource manifests
with open("BP/manifest.json", "w") as file:
    json.dump(bp_manifest, file, indent=4)
with open("RP/manifest.json", "w") as file:
    json.dump(rp_manifest, file, indent=4)

File: tools/process_manifest.mjs
import fs from "fs";

function processJsonElement(element, bpElement, rpElement, options = {}) {
    function process(key, value) {
        if (Array.isArray(value)) {
            bpElement[key] = [];
            rpElement[key] = [];
            processJsonElement(value, bpElement[key], rpElement[key], options);
        } else if (typeof value === "object" && value !== null) {
            bpElement[key] = {};
            rpElement[key] = {};
            processJsonElement(value, bpElement[key], rpElement[key], options);
        } else {
            if (Array.isArray(bpElement)) {
                bpElement.push(value);
                rpElement.push(value);
            } else {
                bpElement[key] = value;
                rpElement[key] = value;
            }
        }
    }

    if (Array.isArray(element)) {
        for (let i = 0; i < element.length; i++) {
            process(i, element[i]);
        }
    } else if (typeof element === "object" && element !== null) {
        for (const [key, value] of Object.entries(element)) {
            if (key.startsWith("bp_")) {
                const bpSpecialPrefixes = [
                    { prefix: "bp_gametest_", condition: options.debugMode },
                    { prefix: "bp_server_", condition: options.isServer },
                    { prefix: "bp_editor_", condition: options.isEditor },
                ];
                let handled = false;
                for (const { prefix, condition } of bpSpecialPrefixes) {
                    if (key.startsWith(prefix)) {
                        if (condition) {
                            const sliceLen = prefix.length;
                            const subKey = key.slice(sliceLen);
                            const sub = bpElement[subKey];
                            if (Array.isArray(sub)) bpElement[subKey] = sub.concat(value);
                            else if (typeof sub === "object" && sub !== null) bpElement[subKey] = { ...sub, ...value };
                        }
                        handled = true;
                        break;
                    }
                }
                if (!handled) bpElement[key.slice(3)] = value;
            } else if (key.startsWith("rp_")) {
                rpElement[key.slice(3)] = value;
            } else {
                process(key, value);
            }
        }
    }
}

function processManifest(debugMode, isServer, isEditor) {
    const bp_manifest = {};
    const rp_manifest = {};

    // load base manifest
    const manifest = JSON.parse(fs.readFileSync("mc_manifest.json", "utf8"));
    processJsonElement(manifest, bp_manifest, rp_manifest, { debugMode, isServer, isEditor });

    let version = manifest.header.version;
    bp_manifest.header.name += " " + (version.join ? version.join(".") : version);
    rp_manifest.header.name += " " + (version.join ? version.join(".") : version);

    if (typeof version !== "string") version = version.slice(0, 3);
    bp_manifest.header.version = version;
    rp_manifest.header.version = version;

    if (!bp_manifest.dependencies) bp_manifest.dependencies = [];
    bp_manifest.dependencies.push({
        uuid: rp_manifest.header.uuid,
        version: rp_manifest.header.version,
    });

    if (!rp_manifest.dependencies) rp_manifest.dependencies = [];
    rp_manifest.dependencies.push({
        uuid: bp_manifest.header.uuid,
        version: bp_manifest.header.version,
    });

    if (debugMode) {
        bp_manifest.header.name += " [DEBUG]";
        rp_manifest.header.name += " [DEBUG]";
    }

    // export behaviour and resource manifests
    fs.mkdirSync("BP", { recursive: true });
    fs.mkdirSync("RP", { recursive: true });
    fs.writeFileSync("BP/manifest.json", JSON.stringify(bp_manifest, undefined, 4));
    fs.writeFileSync("RP/manifest.json", JSON.stringify(rp_manifest, undefined, 4));
}

export default function (args) {
    const debugMode = args.gametest || args.target === "debug";
    const isServer = args.target === "server" || args.target === "server";
    const isEditor = args.editor === true;
    processManifest(debugMode, isServer, isEditor);

    if (args.watch) {
        fs.watch("mc_manifest.json", { persistent: true, recursive: false }, (eventType, filename) => {
            if (filename === "mc_manifest.json") processManifest(debugMode, isServer, isEditor);
        });
    }
}
